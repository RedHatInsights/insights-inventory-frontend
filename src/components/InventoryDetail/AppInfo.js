/* eslint-disable camelcase */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { useLocation } from 'react-router-dom';
import { useStore, useSelector } from 'react-redux';
import { Skeleton, SkeletonSize } from '@redhat-cloud-services/frontend-components/Skeleton';

/**
 * Small component that just renders active detail with some specific class.
 * This component detail is accessed from redux if no component found `missing component` is displayed.
 * @param {*} props `componentsMapper` if you want to pass different components list.
 */
const AppInfo = ({ componentMapper: Cmp, activeApp }) => {
    const store = useStore();
    const loaded = useSelector(({ entityDetails }) => entityDetails?.loaded);
    const entity = useSelector(({ entityDetails }) => entityDetails?.entity);

    if (loaded === true && !entity) {
        return null;
    }

    return (
        <Fragment>
            {
                loaded ? activeApp && (
                    <div className={ `ins-active-app-${activeApp?.name}` }>
                        { Cmp ?
                            <Cmp
                                store={store}
                                inventoryId={entity?.id}
                                appName={activeApp?.name}
                            /> :
                            'missing component'}
                    </div>
                ) : <Skeleton size={ SkeletonSize.md } />
            }
        </Fragment>
    );
};

AppInfo.propTypes = {
    componentMapper: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
    activeApp: PropTypes.shape({
        title: PropTypes.node,
        name: PropTypes.string,
        pageId: PropTypes.string
    })
};

/**
 * This component can be removed once all apps migrate to componentMapper and activeApp.
 * @param { componentMapper, activeApp } props.
 * @returns AppInfo component
 */
const AppInfoWrapper = ({ componentMapper, activeApp, ...props }) => {
    const { search } = useLocation();
    const searchParams = new URLSearchParams(search);
    if (!componentMapper || !activeApp) {
        console.warn('Please pass componentMapper and activeApp. We will be deprecating the old store controls');
    }

    const currApp = useSelector(({ entityDetails }) => {
        const activeItem = searchParams.get('appName') || entityDetails?.activeApp?.appName;
        return entityDetails?.activeApps?.find?.(item => item?.name === activeItem) || entityDetails?.activeApps?.[0];
    });

    const currComponent = componentMapper || (activeApp || currApp)?.component;

    return <AppInfo componentMapper={currComponent} activeApp={activeApp || currApp} {...props} />;
};

AppInfoWrapper.propTypes = AppInfo.propTypes;

// TODO: Remove once all apps send `componentMapper` and `activeApp` and use directly AppInfo
const AppInfoCmp = (props) => (props.componentMapper && props.activeApp) ? <AppInfo {...props} /> : <AppInfoWrapper {...props} />;

AppInfoCmp.propTypes = AppInfo.propTypes;

export default AppInfoCmp;
