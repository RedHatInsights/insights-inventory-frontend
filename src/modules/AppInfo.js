import React from 'react';
import AsyncInventory from './AsyncInventory';
import { AppInfo as AppInfoCmp } from '../components/InventoryDetail';

const AppInfo = (props) => <AsyncInventory {...props} component={AppInfoCmp} />;

export default AppInfo;
