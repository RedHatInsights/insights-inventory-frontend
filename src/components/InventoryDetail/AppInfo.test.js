/* eslint-disable camelcase */
import React from 'react';
import AppInfo from './AppInfo';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { createPromise as promiseMiddleware } from 'redux-promise-middleware';
import { NotConnected } from '@redhat-cloud-services/frontend-components/NotConnected';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => ({
        pathname: '/some/path'
    })
}));

let initialState = {};
let mockStore;
let wrapper;
beforeEach(() => {
    mockStore = configureStore([promiseMiddleware()]);
    initialState = {
        entityDetails: {
            loaded: true,
            entity: {
                insights_id: '0253a9b9-f9fb-4ba5-a62e-28cbb3cb16db',
                stale_timestamp: '2260-01-01T00:00:00+00:00',
                stale_warning_timestamp: '2260-01-08T00:00:00+00:00',
                culled_timestamp: '2260-01-15T00:00:00+00:00',
                created: '2022-09-29T10:17:38.158029+00:00',
                updated: '2022-11-14T21:40:08.619998+00:00',
                system_profile: {
                    operating_system: {
                        major: 9,
                        minor: 0,
                        name: 'RHEL'
                    }
                },
                tags: [],
                per_reporter_staleness: {
                    puptoo: {
                        check_in_succeeded: true,
                        stale_timestamp: '2260-01-01T00:00:00+00:00',
                        last_check_in: '2022-11-11T11:29:50.154804+00:00'
                    }
                }
            },
            isToggleOpened: true
        }
    };

    const store = mockStore(initialState);
    wrapper = mount(<Provider store={store}>
        <AppInfo />
    </Provider>);
});

describe('AppInfo', () => {
    it('should render correctly', () => {
        expect(wrapper.find(NotConnected)).toMatchSnapshot();
    });
    it('should render NotConnected component', () => {
        const store = mockStore(initialState);
        const tempWrapper = mount(<Provider store={store}>
            <AppInfo notConnected={true}/>
        </Provider>);

        expect(tempWrapper.find(NotConnected)).toHaveLength(1);
    });

});
