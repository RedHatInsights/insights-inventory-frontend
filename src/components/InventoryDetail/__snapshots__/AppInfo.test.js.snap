// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppInfo should render correctly 1`] = `ReactWrapper {}`;
